### PR TITLE
Custom cluster feature creation function

### PR DESCRIPTION
**Pull request by Razi91**
Adds `hydrate` function option into Cluster constructor. It allows to for example fill cluster's feature with attributes, set custom style etc.

I'm not sure if `hydrate` is right name here. It's called after Feature is created, before `features` attribute is set.

I'm not able to run tests locally (due to https://github.com/openlayers/openlayers/issues/12486)

example usage:
```ts
      const source = new Cluster({
        source: new VectorSource({
          features: [feature1, feature2],
        }),
        hydrate: (clusterFeature, features) => {
          clusterFeature.set('sum', features.reduce((acc, ft) => acc + ft.get('value'), 0);
          clusterFeature.set('hasAllValidItems', features.every((ft) => acc + ft.get('isValid'));
        },
      });
```

It might be useful when you receive events elsewhere and you want to have for example the same attribute in every feature on map, whether it's clustered or not

// EDIT
Changed to  `createClusterFeature: (clusterPoint: Point, featuresInCluster: Feature[]): Feature`, it will be more elastic. Now inserting features is optional and can be done under different attribute name. Also we can use different Feature implementation (if anyone needs it), reuse etc.

### CODE REVIEW

**Comment by mike-000** [comment]
`aggregate` might be a better name as it is intended for property value aggregation

**Comment by ahocevar** [comment]
What about allowing the user complete control over creation of the cluster feature, with an createClusterFeature option? The current behavior (default) would be
```js
  createClusterFeature: (clusterPoint, featuresInCluster) => {
    return new Feature({
      geometry: clusterPoint,
      features: featuresInCluster
    });
  }
```

**Comment by Razi91** [comment]
Sounds good @ahocevar , AFAIR `'features'` attribute is not used anywhere in OpenLayers itself, it's just for application developers. That way we'll have full control over how this Feature will look like.

**Comment by Razi91** [comment]
Changed and rebased to new version. Should I squash those commits?

**Review by ahocevar — COMMENTED** [review]
Please add better documentation. It would be a good idea to add the default code as snippet to the docs, so users get a clear picture of what this function could look like.

**Comment by Razi91** [comment]
Added better documentation, I was considering also adding an example or modify one, but I don't have any idea right now about it :(

File: src/ol/source/Cluster.js [hunk 1]
```diff
@@ -40,6 +40,18 @@ import {getUid} from '../util.js';
  * ```
  * See {@link module:ol/geom/Polygon~Polygon#getInteriorPoint} for a way to get a cluster
  * calculation point for polygons.
+ * @property {function(Point, Array<Feature>):Feature} [createClusterFeature]
+ * Function that takes cluster's ceter {@link module:ol/geom/Point} and array
+ * of {@link module:ol/Feature} included in this cluster, must return
```
Comment (ahocevar) [inline src/ol/source/Cluster.js:45]:
```suggestion
 * Function that takes the cluster's center {@link module:ol/geom/Point} and an array
 * of {@link module:ol/Feature} included in this cluster. Must return a
```

File: src/ol/source/Cluster.js [hunk 2]
```diff
@@ -180,6 +198,28 @@ class Cluster extends VectorSource {
     this.updateDistance(this.distance, minDistance);
   }
 
+  /**
+   * Current function used to create cluster features
+   * @return {function(Point, Array<Feature>):Feature | undefined} cluster creation function
+   * @api
+   */
+  getCreateClusterFeature() {
+    return this.createClusterFeature;
```
Comment (ahocevar) [inline src/ol/source/Cluster.js:207]: If you provide this getter, it should also return the default if no custom function is configured.

Comment (Razi91) [reply src/ol/source/Cluster.js:207]: I'm not sure if it should return default. This way we can check if any is provided, or default is used. If we return the default, we'll need to check if it's the one we want or not (or if it's default)

Comment (ahocevar) [reply src/ol/source/Cluster.js:207]: If you export the default function, you can easily check if it's the default:
```js
if (cluster.getCreateClusterFeature() === defaultCreateClusterFeature)
```
But I'd be inclined to not have a `getCreateClusterFeature()` getter at all, and maybe not even a setter. What would be a use case for needing those? It's cheap to create a new cluster source.

**Comment by Razi91** [comment]
Ok, removed getter/setter, fixed docs.

**Review by ahocevar — APPROVED** [review]
Thanks @Razi91, this looks great now!

**Pull request merged**